alf/algorithms/containers.py

```diff
@@ -0,0 +1,622 @@
+            priority = add_ignore_empty(loss_info.priority, priority)
+
+        return LossInfo(
+            loss=loss, scalar_loss=scalar_loss, extra=extra, priority=priority)
+
+    def preprocess_experience(self, root_inputs, rollout_info, batch_info):
+        """Call the preprocess_experience of each sub-algorithm."""
+        new_infos = {}
+        for name, alg in self._algs.items():
+            root_inputs, info = alg.preprocess_experience(
+                root_inputs, rollout_info[name], batch_info)
+            new_infos[name] = info
+
+        return root_inputs, new_infos
+
+    def after_update(self, root_inputs, info):
+        """Call after_update of each sub-algorithm."""
+        for name, alg in self._algs.items():
+            alg.after_update(root_inputs, info[name])
+
+    def after_train_iter(self, root_inputs, rollout_info):
+        """Call after_train_iter of each sub-algorithm."""
+        for name, alg in self._algs.items():
+            alg.after_train_iter(root_inputs, rollout_info[name])
+
+
+def SequentialAlg(*modules,
+                  output='',
+                  is_on_policy=None,
+                  name="SequentialAlg",
+                  **named_modules):
+    """Compose Algorithms Networks sequentially as a new Algorithm.
+
+    All the modules provided through ``modules`` and ``named_modules`` are calculated
+    sequentially in the same order as they appear in the call to ``SequentialAlg``.
+    By default, each module takes the output of the previous module as its input
+    (or the input to the Sequential if it is the first module), and the output of
+    the last module is the output of the ``SequentialAlg``. Note that the output
+    of a module means differently depending on the type of the module:
+
+    * Algorithm: ``AlgStep.output`` field from ``predict_step``, ``rollout_step``
+        or ``train_step``
+    * Network: the first element of the tuple returned from ``forward()``
+    * torch.nn.Module or Callable: the return value of the Callable.
+
+    In addition to using the output of the previous module as input,
+    ``SequentialAlg`` also allow using other output, state or info from previous
+    module as the input to a module. To do this,
+    one can pass a tuple of (nested_str, module) instead of module as an argument
+    to ``SequentialAlg``. With this, the inputs to the module will be obtained using
+    ``get_nested_field(named_results, nested_str)``, where ``named_results``
+    is a dictionary containing the inputs to ``SequentialAlg`` and all the results calulcated
+    by previous modules. More specifically, ``named_results['input']`` is the
+    inputs to this algorithm. ``named_results['a']`` is the output of the module
+    named 'a'. ``named_results['info']['a']`` is the info output of the algorithm
+    named 'a'.  And ``named_results['state']['a']`` is state output of the
+    algorithm/network named 'a'.
```

Comment:
Seems still missing the explanation for loss computation. Is it always use the 'loss=' field when retrieving the loss?